declarations/index.d.ts

```diff
@@ -1,22 +1,107 @@
 /// <reference types="node" />
 import { EventEmitter } from 'events';
-export interface IAbstractIO {
-    isReady: boolean;
-    readonly MODES: {
-        INPUT: number;
-        OUTPUT: number;
-        ANALOG: number;
-        PWM: number;
-        SERVO: number;
-    };
+export declare enum Mode {
+    INPUT = 0,
+    OUTPUT = 1,
+    ANALOG = 2,
+    PWM = 3,
+    SERVO = 4,
+    UNKOWN = 99
+}
+export declare enum Value {
+    HIGH = 1,
+    LOW = 0
+}
+export interface IPinConfiguration {
+    supportedModes: Mode[];
+    mode: Mode;
+    value: Value;
+    reporting: 0 | 1;
+    analogChannel: number;
+}
+export interface IPingReadSettings {
+    pin: number | string;
+    value?: Value;
+    pulseOut?: number;
+}
+export interface II2CConfig {
+    address: number;
+    bus?: number | string;
+    port?: number;
+    delay?: number;
+}
```

Comment:
We may need data, parity and stop bits. Looking into it...

Reply:
Does I2C have those too? I only recall UART having them, but I could be wrong.

Reply:
No, I put my comment in the wrong place. Should be on ISerialConfig. Sorry.

Reply:
Looking through the Johnny-Five codebase, we only call `serialConfig` with the options already included (no data, parity, or stop bits). The firmata.js API docs seem to say the same thing, as does the firmata protocol docs themselves (I think, they're kinda hard to read).

It does seem odd though that data, parity, and stop bits aren't included though.

My inclination is to leave them out for now in this PR.

Reply:
That makes sense. I did some diggin and found where @soundanalagous chose to punt on those extra params. We can always add them in if we find a need down the road.

Reply:
🤔 good point, these properties are easy enough to add, and since they're optional it's backwards compatible with existing plugins. It should be easy to add to Raspi IO once this is done and Raspi IO is refactored on top of it.

Reply:
Note: re baud rate. The Uno only has one UART so it defaults to software serial which gets kinda sketchy around 57,600

Reply:
I don't think that'll cause any issues from the perspective of this base class. I'd categorize not being able to support a specific baud rate the same as not being able to support a specific method...the IO plugin that derives from Abstract IO should throw an exception.

Reply:
I don't think we want to define a constructor, since it varies from IO Plugin to IO Plugin. Raspi IO and Spark IO alone have really different constructors.

I would claim that anything baudrate related for talking to firmata devices should be in that IO Plugin constructor (aka the firmata.js constructor), not the base class here.

Reply:
So how's everyone feeling on this conversation thread? Do we all feel that serial is properly covered now and I can mark this as resolved, or is there some more work to do?